vdirsyncer/sync.py

```diff
@@ -99,107 +102,328 @@ def to_ident_conflict(self, storage):
+    def insert_ident_b(self, ident, props):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def update_ident_a(self, ident, props):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def update_ident_b(self, ident, props):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def remove_ident(self, ident):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_a(self, ident):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_b(self, ident):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_new_a(self, ident):
+        raise NotImplementedError()
 
-        self._href_to_status_a = dict((meta['href'], (ident, meta))
-                                      for ident, (meta, _)
-                                      in self._ident_to_props.items())
+    @abc.abstractmethod
+    def get_new_b(self, ident):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def iter_old(self):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def iter_new(self):
+        raise NotImplementedError()
 
-        self._href_to_status_b = dict((meta['href'], (ident, meta))
-                                      for ident, (_, meta)
-                                      in self._ident_to_props.items())
+    @abc.abstractmethod
+    def get_by_href_a(self, href, default=(None, None)):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_by_href_b(self, href, default=(None, None)):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def rollback(self, ident):
+        raise NotImplementedError()
+
+
+class _SqliteStatus(_StatusBase):
+    SCHEMA_VERSION = 1
+
+    def __init__(self, path):
+        self._path = path
+        self._conn = sqlite3.connect(path)
+        self._conn.row_factory = sqlite3.Row
+        self._c = None
+
+        if self._is_latest_version():
+            return
+
+        self._conn.execute('DROP TABLE IF EXISTS status')
+        self._conn.execute('DROP TABLE IF EXISTS new_status')
+        self._conn.execute('DROP TABLE IF EXISTS meta')
+
+        self._conn.execute('''CREATE TABLE meta (
+            "version" INTEGER PRIMARY KEY
+        ); ''')
+        self._conn.execute('INSERT INTO meta (version) VALUES (?)',
+                           (self.SCHEMA_VERSION,))
+
+        self._conn.execute('''CREATE TABLE status (
+            "ident" TEXT PRIMARY KEY NOT NULL,
+            "href_a" TEXT NOT NULL,
+            "href_b" TEXT NOT NULL,
+            "hash_a" TEXT NOT NULL,
+            "hash_b" TEXT NOT NULL,
+            "etag_a" TEXT NOT NULL,
+            "etag_b" TEXT NOT NULL
+        ); ''')
+        self._conn.execute('CREATE UNIQUE INDEX by_href_a ON status(href_a)')
+        self._conn.execute('CREATE UNIQUE INDEX by_href_b ON status(href_b)')
+
+        # We cannot add NOT NULL here because data is first fetched for the
```

Comment:
> The NULLs really need to be there for new_status because while the program is running, NULLs can occur

I was actually thinking that you could only have *one* status table and not two status tables, since you can update (and persist) rows at once now in the new schema. Am I mistaken here?

Either way, it's probably out-of-scope for this PR and something that warrants a separate one.